src/formats/XYZ.cpp

```diff
@@ -78,19 +103,396 @@ optional<uint64_t> XYZFormat::forward() {
         // We could not read an integer, so give up here
         return nullopt;
     }
-    static int step = 0;
-    step++;
 
     for (size_t i=0; i<natoms + 1; i++) {
         if (file_.eof()) {
             throw format_error(
                 "XYZ format: not enough lines at step {} (expected {}, got {})",
-                step, natoms + 2, i + 1
+                forward_step_, natoms + 2, i + 1
             );
         }
 
         file_.readline();
     }
 
+    forward_step_++;
+
     return position;
 }
+
+/*****************************************************************************/
+/* End of the main XYZ format implementation. The code below implements the  */
+/* extended XYZ format/convention. This format is 100% backward compatible   */
+/* with normal XYZ. It adds data in two places: after the atomic positions   */
+/* and in the comment line. The comment line contains key/values pairs       */
+/* describing frame-level properties, as well as two special keys: Lattice   */
+/* stores the unit cell matrix, and 'Properties' describe the name and type  */
+/* of any additional atom-level data.                                        */
+/*****************************************************************************/
+
+static bool contains_space(string_view s) {
+    for (auto c: s) {
+        if (is_whitespace(c)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+std::string write_extended_comment_line(const Frame& frame) {
+    std::string result;
+    // TODO: store more properties? Atom-level properties are a good idea, but
+    // we don't have guarantee that they exist for all atoms. Residue
+    // information could also be nice (resid/resname).
+    result += "Properties=species:S:1:pos:R:3";
+
+    // support for Lattice
+    if (frame.cell().shape() != UnitCell::INFINITE) {
+        auto matrix = frame.cell().matrix();
+        // set small elements to 0
+        for (size_t i=0; i<3; i++) {
+            for (size_t j=0; j<3; j++) {
+                if (matrix[i][j] < 1e-12) {
+                    matrix[i][j] = 0;
+                }
+            }
+        }
+        result += fmt::format(" Lattice=\"{:g} {:g} {:g} {:g} {:g} {:g} {:g} {:g} {:g}\"",
+            matrix[0][0], matrix[1][0], matrix[2][0],
+            matrix[0][1], matrix[1][1], matrix[2][1],
+            matrix[0][2], matrix[1][2], matrix[2][2]
+        );
+    }
+
+    auto sorted_properties = std::map<string_view, Property>{frame.properties().begin(), frame.properties().end()};
+    // support for generic frame properties
+    for (const auto& it: sorted_properties) {
+        if (contains_space(it.first)) {
+            warning("Extended XYZ", "frame property '{}' contains spaces, skipping", it.first);
```

Comment:
Are we sure, that spaces in keys are explicitly invalid? Couldn't find anything about this in the spec and QUIP seems to [allow](https://github.com/libAtoms/QUIP/blob/10245cffc56cf5d0a3de6e5d3475386d7e4f3c80/src/libAtoms/xyz.c#L701-L742) quoted keys (at least for reading; based on their comments, so I'm not fully sure about that).

Reply:
I'll try a file with quoted key with QUIP to check this. My reading of the documentation made me think quotes were only allowed in values, but you are right that they are not explicitly forbidden for keys ...

Reply:
ASE reads quoted keys with spaces fine, so I'll have to update this

Reply:
This is more complex than I though, I've asked ASE maintainers about this: https://gitlab.com/ase/ase/issues/474

Reply:
The implementation should be working now, but we are still missing a standard other than 'whatever ASE does'

Reply:
I guess out best bet is to do [whatever ase does](https://gitlab.com/ase/ase/-/blob/master/ase/io/extxyz.py#L604-713) for now. The current state of this PR is already quite well and might be useful to some users (e.g. for me personally =)). Figuring out a comprehensive standard could be part of #280.